docs/type.rst

```diff
@@ -46,54 +71,128 @@ Currently, supported basic types in Taichi are
     (OK: supported, EXT: require extension, N/A: not available)
 
 
-Boolean types should be represented using ``ti.i32``.
+.. note::
+
+    Boolean types are represented using ``ti.i32``.
+
 
-Binary operations on different types will give you a promoted type, following the C programming language, e.g.
 
-- ``i32 + f32 = f32``
-- ``f32 + f64 = f64``
-- ``i32 + i64 = i64``
+Type promotion
+--------------
+
+Binary operations on different types will give you a promoted type, following the C programming language convention, e.g.:
+
+- ``i32 + f32 = f32`` (integer + float = float)
+- ``i32 + i64 = i64`` (less-bits + more-bits = more-bits)
+
+Basically it will try to choose the more precise type to contain the result value.
 
 
 .. _default_precisions:
 
 Default precisions
 ------------------
 
-By default, numerical literals have 32-bit precisions.
+By default, all numerical literals have 32-bit precisions.
 For example, ``42`` has type ``ti.i32`` and ``3.14`` has type ``ti.f32``.
+
 Default integer and float-point precisions (``default_ip`` and ``default_fp``) can be specified when initializing Taichi:
 
 .. code-block:: python
 
-    ti.init(..., default_fp=ti.f32)
-    ti.init(..., default_fp=ti.f64)
+    ti.init(default_fp=ti.f32)
+    ti.init(default_fp=ti.f64)
+
+    ti.init(default_ip=ti.i32)
+    ti.init(default_ip=ti.i64)
+
+
+Also note that you may use ``float`` or ``int`` in type definitions as aliases
+for default precisions, e.g.:
+
+.. code-block:: python
+
+    ti.init(default_ip=ti.i64, default_fp=ti.f32)
+
+    x = ti.var(float, 5)
+    y = ti.var(int, 5)
+    # is equivalent to:
+    x = ti.var(ti.f32, 5)
+    y = ti.var(ti.i64, 5)
+
+    def func(a: float) -> int:
+        ...
+
+    # is equivalent to:
+    def func(a: ti.f32) -> ti.i64:
+        ...
 
-    ti.init(..., default_ip=ti.i32)
-    ti.init(..., default_ip=ti.i64)
 
 
 Type casts
 ----------
 
-Use ``ti.cast`` to cast scalar values.
+Implicit casts
+**************
+
+WARNING: The type of a variable is **determinated on it's initialization**.
+
+When a *wide* variable is assigned with a *narrow* type, it will be
+implicitly promoted to the *wide* type and no warning will be raised:
+
+.. code-block:: python
+
+    a = 1.7
+    a = 1
+    print(a)  # 1.0
+
+When a *narrow* variable is assigned with a *wide* type, it will be
+implicitly casted into the *narrow* type and Taichi will raise a warning:
```

Comment:
```suggestion
When a *high-precision* variable is assigned to a *low-precision* type, it will be
implicitly down-cast into the *low-precision* type and Taichi will raise a warning:
```